securesystemslib/kms_signer.py

```diff
@@ -0,0 +1,104 @@
+"""KMS Signer implementations"""
+
+import logging
+from typing import Any, Dict
+
+import securesystemslib.hash as sslib_hash
+from securesystemslib import exceptions, formats
+from securesystemslib.signer import Signature, Signer
+
+logger = logging.getLogger(__name__)
+
+NO_GCP_MSG = "Google Cloud keys ('gcpkms://') require google-cloud-kms library"
+GCP = True
+try:
+    from google.cloud import kms
+except ImportError:
+    GCP = False
+
+
+class KMSSigner(Signer):  # pylint: disable=W0223
+    """An abstract KMS Signer
+
+    KMSSigner signs using a Key Management Service. The specific KMS implementation
+    is chosen using the "keyref" string which identifies the KMS private key.
+    """
+
+    @staticmethod
+    def from_keyref(keyref: str, public_key: Dict[str, Any]):
+        """Returns a concrete KMSSigner implementation
+
+        Args:
+            keyref: A KMS-specific URI that identifies the signing key
+            public_key: Public key metadata conforming to PUBLIC_KEY_SCHEMA
+        """
+
+        formats.PUBLIC_KEY_SCHEMA.check_match(public_key)
+
+        if keyref.startswith(GCPSigner.KEYREF_SCHEME):
+            if not GCP:
+                raise exceptions.UnsupportedLibraryError(
+                    f"google-cloud-kms library required for KMS key {keyref}",
+                )
+            return GCPSigner(keyref, public_key)
+
+        raise ValueError(f"Unrecognized keyref {keyref}")
+
+    @staticmethod
+    def _hash_algorithm_for_public_key(key: Dict[str, Any]) -> str:
+        """Returns the payload hash algorithm for given public key dict"""
+
+        if key["keytype"] == "rsa":
+            # hash algorithm is encoded as last scheme portion
+            return key["scheme"].split("-")[-1]
+        if key["keytype"] in [
+            "ecdsa",
+            "ecdsa-sha2-nistp256",
+            "ecdsa-sha2-nistp384",
+        ]:
+            # nistp256 uses sha-256, nistp384 uses sha-384
+            bits = key["scheme"].split("-nistp")[-1]
+            return f"sha{bits}"
+        if key["keytype"] == "ed25519" and key["scheme"] == "ed25519":
+            # specification default is sha-512
+            return "sha512"
```

Comment:
I'm not sure where I got this impression from...

Reply:
Hmm, the spec defines sha512 and the identity function as the prehash options. Using identity (i.e. no hashing) in this case would be silly...

However, spec uses "ed25519" to mean the identity prehashed scheme, and "ed25519ph" as the SHA512 prehashed scheme: It's entrirely possible that our current implementation does the former and then we would have to add a new scheme to the ed25519 keytype.

Considering that GCP KMS does not even support ed25519, I think I'll just remove this if-clause